text/3861-iter-macro.md

```diff
@@ -0,0 +1,496 @@
+For iterator closures, these are computed the same as for any other closure.
+For the corresponding iterator, auto traits are computed based on the values that must be saved across a `yield` point, similar to how auto traits for an `async` block are computed based on the values that are saved across `await` points.
+
+It is not possible to hold a borrow of a value local to the iterator across a `yield` point.
+Similarly, it is not possible to `yield` a reference to an iterator-local value.
+Note that holding and yielding references that completely outlive the iterator closure (such as a reference captured from the environment) are allowed.
+
+# Drawbacks
+
+- Iterator closures have some limitations around self-borrows and lending.
+- The macro-based syntax looks less built-in than if it were actually built-in.
+- Does not exactly support the plain iterator block use case. You always have to create and invoke a closure.
+
+# Rationale and alternatives
+
+This incremental step forward will help us make continued progress towards more advanced generators.
+If we do not take this step, we expect we will see little further development in generators and users will remain frustrated that they do not have convenient syntax for creating iterators.
+This is a feature that is often missed by Rust programmers.
+
+## Why Iterator Closures?
+
+It might seem more obvious to have `iter!` evaluate directly to an `impl Iterator` with no intermediate closure step.
+We instead recommend returning an iterator closure.
+This is largely as a result of what we have learned from our experience with `async`.
+
+Having a two step process between creating the iterator-like object and beginning iteration allows us to support scenarios such as where the result of `iter!` is `Send` but the iterator is no longer `Send` once iteration starts.
+See Yosh Wuyts' [The Gen Auto-Trait Problem] for more details.
+In `async`, we've had a lot of discussion about using `IntoFuture` for this two stage process but decided that it is better represented through async closures.
+For iterators and generators, we'd like to set the same precedent from the beginning.
+
+[The Gen Auto-Trait Problem]: https://blog.yoshuawuyts.com/gen-auto-trait-problem/
+
+Second, having convenient syntax for creating inline iterators will create an incentive to create more powerful combinators.
+With async, people very quickly started writing functions that took arguments with types like `impl FnOnce() -> F where F: Future`.
+Despite the clear desire to write this, these never worked particularly well until we had proper support for async closures.
+Still, this created an ecosystem hazard, as we wanted what Rust supported to be broadly compatible with how the ecosystem had already been experimenting.
+Again, using what we learned from async, we have the chance to do the right thing from the beginning with iterator closures.
```

Comment:
I'm not really sure what this paragraph is saying. The second and third sentence imply you want people to write combinators with `impl IterFn()` syntax, but that's not stable. Then the example below talks about using combinators with a hypothetical `then` method on iterators, which isn't part of std. So I'm not really sure how we're encouraging people to use combinators at all as part of this RFC.